packages/react/src/components/form/initial/controls.tsx

```diff
@@ -172,16 +173,33 @@ const FormInput = ({ children }: FormInputProps) => {
   const { lastHandle } = useInternalFormContext();
   const { factors, text } = useConfiguration();
 
+  const [showPasskeys, setShowPasskeys] = useState<boolean | null>(null);
+  useEffect(() => {
+    (async () => {
+      const isSupported = await runtimeSupportsPasskeys;
+      setShowPasskeys(isSupported);
+    })();
+  }, []);
+
   // @ts-expect-error TODO fix inference
   const nonOidcFactors: FactorNonOIDC[] = useMemo(
-    () => factors.filter((f) => isFactorNonOidc(f)),
-    [factors]
+    () =>
+      factors.filter((f) => {
+        if (f.method === "webauthn" && showPasskeys === false) return false;
+        return isFactorNonOidc(f);
+      }),
+    [factors, showPasskeys]
   );
 
   const handleTypes = useMemo(() => {
     return getHandleTypes(factors);
   }, [factors]);
 
+  const passkeySupportDetectionLoading = showPasskeys === null;
+  if (passkeySupportDetectionLoading) {
+    return null;
```

Comment:
Ideally this would be a skeleton or some sort of placeholder, so we can avoid layout shifts.

Reply:
Agreed